cou/steps/ceph.py

```diff
@@ -0,0 +1,279 @@
+    the model.
+
+    :param model: The juju model to work with
+    :type model: Model
+    :param enable: True to set noout, False to unset noout
+    :type enable: bool
+    """
+    try:
+        ceph_mon_apps = await _get_applications(model)
+    except ApplicationNotFound as e:
+        logger.warning("%s", str(e))
+        logger.warning("Skip changing 'noout', because there's no ceph-mon applications.")
+        return
+
+    for ceph_mon_app in ceph_mon_apps:
+        try:
+            ceph_mon_unit_name = await _get_unit_name(ceph_mon_app)
+        except UnitNotFound as e:
+            logger.warning("%s", str(e))
+            logger.warning(
+                "Skip changing 'noout', because there's no %s units.", ceph_mon_app.name
+            )
+            continue
+
+        await model.run_action(
+            ceph_mon_unit_name, f"{'set' if enable else 'unset'}-noout", raise_on_failure=True
+        )
+
+
+async def get_osd_noout_state(model: Model, unit_name: str) -> bool:
+    """Check if noout is set on ceph cluster or not.
+
+    :param model: The juju model to work with
+    :type model: Model
+    :param unit_name: The name of the unit
+    :type unit_name: str
+    :return: True if noout is set, otherwise False
+    :type: bool
+    """
+    results = await model.run_on_unit(unit_name, "ceph osd dump -f json")
+    return "noout" in json.loads(results["stdout"].strip()).get("flags_set", [])
+
+
+async def assert_osd_noout_state(model: Model, state: bool) -> None:
+    """Assert ceph cluster is set (state=True) or unset (state=False).
+
+    Note this will assert 'noout' flag is in the desired state for all ceph
+    clusters present in the model.
+
+    :param model: The juju model to work with
+    :type model: Model
+    :param state: True noout is set, otherwise False
+    :type state: bool
+    :raise ApplicationError: if noout is not in desired state
+    """
+    try:
+        ceph_mon_apps = await _get_applications(model)
+    except ApplicationNotFound as e:
+        logger.warning("%s", str(e))
+        logger.warning("Skip verifying 'noout', because there's no ceph-mon applications.")
+        return
+
+    error = False
+    for ceph_mon_app in ceph_mon_apps:
+        try:
+            ceph_mon_unit_name = await _get_unit_name(ceph_mon_app)
+        except UnitNotFound as e:
+            logger.warning("%s", str(e))
+            logger.warning(
+                "Skip verifying 'noout', because there's no %s units.", ceph_mon_app.name
+            )
+            continue
+
+        if await get_osd_noout_state(model, ceph_mon_unit_name) is not state:
+            error = True
+            logger.error(
+                "'noout' is expected to be %s for %s",
+                "set" if state else "unset",
+                ceph_mon_app.name,
+            )
+
+    if error:
+        raise ApplicationError("Ceph cluster's 'noout' is not in expected state.")
```

Comment:
I think it's necessary, this is checking if any ceph cluster has incorrect `noout` state. Putting it at the end will always raise `ApplicationError`

Reply:
What I'm trying to say is that this flag error is used just in the scenario above. Why not delete this flag and raise the error directly if the noout state is not in the expected state?

Reply:
For multiple ceph clusters, multiple ceph clusters can be in unexpected state, so it will check and report all (`logger.error`) errors, and finally raise the errors to stop COU. The feedback can help operator to fix issues all at once, and not by trial and error

Reply:
I see...
I think it would make more explicit if the `ApplicationError` showed the applications name that failed, but I'm OK with the approach

Reply:
there are error logs showing which app has wrong state, and that's why the raise only show overall status